core-api/src/main/java/com/optimizely/ab/Optimizely.java

```diff
@@ -397,18 +401,32 @@ public Boolean isFeatureEnabled(@Nonnull String featureKey,
                     userId,
                     copiedAttributes,
                     featureDecision.variation);
+                decisionSource = featureDecision.decisionSource;
+                sourceVariationKey = featureDecision.variation.getKey();
+                sourceExperimentKey = featureDecision.experiment.getKey();
             } else {
                 logger.info("The user \"{}\" is not included in an experiment for feature \"{}\".",
                     userId, featureKey);
             }
             if (featureDecision.variation.getFeatureEnabled()) {
                 logger.info("Feature \"{}\" is enabled for user \"{}\".", featureKey, userId);
-                return true;
+                featureEnabled = true;
             }
         }
 
+        DecisionNotification decisionNotification = DecisionNotification.newFeatureDecisionNotificationBuilder()
+            .withUserId(userId)
+            .withAttributes(copiedAttributes)
+            .withFeatureKey(featureKey)
+            .withFeatureEnabled(featureEnabled)
+            .withSource(decisionSource)
+            .withExperimentKey(sourceExperimentKey)
+            .withVariationKey(sourceVariationKey)
```

Comment:
May be remove these 2 build options and replace with `withSourceInfo` option which takes `SourceInfo` object.

You can then define 2 implementations `SourceInfo`. 1 for rollout and 1 for feature test and then use a get method on them to get the source info that you build in the builder here. That may be a better way to structure this.

Reply:
Are you suggesting to make methods of get source info in optimizely or in DecisionNotification.java? 
By the way I think current implementation is much simpler and fully utilizing builder pattern.

Reply:
I am suggesting something like:
```
public interface SourceInfo {
  HashMap<String, Object> get()
}

public class FeatureTestSourceInfo implements SourceInfo {
}

public class RolloutSourceInfo implements SourceInfo {
}
```

This may allow us to hopefully add more `SourceInfo` options in the future.

Reply:
Discussed offline. This will happen in a follow up.